crates/bevy_ui/src/accessibility.rs

```diff
@@ -0,0 +1,219 @@
+            ..default()
+        };
+        commands
+            .entity(entity)
+            .insert(AccessibilityNode::from(node));
+    }
+}
+
+fn image_changed(
+    mut commands: Commands,
+    query: Query<(Entity, &GlobalTransform, &Node, &Children), (Changed<UiImage>, Without<Button>)>,
+    texts: Query<&Text>,
+) {
+    for (entity, transform, node, children) in &query {
+        let node = AccessKitNode {
+            role: Role::Image,
+            bounds: Some(calc_bounds(transform, node)),
+            name: calc_name(&texts, children),
+            ..default()
+        };
+        commands
+            .entity(entity)
+            .insert(AccessibilityNode::from(node));
+    }
+}
+
+fn label_changed(
+    mut commands: Commands,
+    query: Query<(Entity, &GlobalTransform, &Node, &Text), Changed<Label>>,
+) {
+    for (entity, transform, node, text) in &query {
+        let values = text
+            .sections
+            .iter()
+            .map(|v| v.value.to_string())
+            .collect::<Vec<String>>();
+        let name = values.join(" ");
+        let bounds = Rect::new(
+            transform.translation().x.into(),
+            transform.translation().y.into(),
+            (transform.translation().x + node.calculated_size.x).into(),
+            (transform.translation().y + node.calculated_size.y).into(),
+        );
+        let node = AccessKitNode {
+            role: Role::LabelText,
+            bounds: Some(bounds),
+            name: Some(name.into_boxed_str()),
+            ..default()
+        };
+        commands
+            .entity(entity)
+            .insert(AccessibilityNode::from(node));
+    }
+}
+
+#[derive(Resource, Default, Deref, DerefMut)]
+struct InteractionCache(HashMap<Entity, Interaction>);
+
+fn interaction_changed(
+    mut cache: ResMut<InteractionCache>,
+    adapters: NonSend<Adapters>,
+    query: Query<(Entity, &Interaction), Changed<Interaction>>,
+) {
+    for (entity, interaction) in &query {
+        if let Some(adapter) = adapters.get_primary_adapter() {
+            if *interaction == Interaction::Hovered {
+                let focus = Some(entity.to_node_id());
+                adapter.update(TreeUpdate { focus, ..default() });
+            } else if let Some(old_interaction) = cache.get(&entity) {
+                if *old_interaction == Interaction::Hovered {
+                    adapter.update(TreeUpdate {
+                        focus: None,
+                        ..default()
+                    });
+                }
+            }
+        }
+        cache.insert(entity, interaction.clone());
+    }
+}
+
+fn interaction_removed(
+    mut cache: ResMut<InteractionCache>,
+    adapters: NonSend<Adapters>,
+    removed: RemovedComponents<Interaction>,
+) {
+    for entity in removed.iter() {
+        if let Some(old_interaction) = cache.get(&entity) {
+            if *old_interaction == Interaction::Hovered {
```

Comment:
```suggestion
        if let Some(&Interaction::Hovered) = cache.get(&entity) {
```
This or `matches!` is probably going to be shorter and more readable.

Alternatively, `let.. else` could reduce nesting here as well.